### DISCUSSION TITLE
renovate-bot close all pull requests without any clear reason

### DISCUSSION THREAD

**Discussion by tmortagne** [renovatebot/renovate#26473] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

None

### Describe the problem

Regularly (not really sure of the frequency but it happens several times in the few months we have been using Renovate) renovate-bot decides to close all the remaining pull requests (old or recent at the exact same tile) it created without about given reason and then react as if we were the ones having closed the issues.

See for example https://github.com/xwiki/xwiki-commons/pull/714

I cannot find any log describing the close, but I can find two logs 31s apart (the time it took for the first one to run) triggered right after the close with reason `requested`, but nobody on our side requested it.

### Relevant debug logs

<details><summary>Logs</summary>

```

DEBUG: syncBranchState() (branch="renovate/slf4j.version")
DEBUG: syncBranchState(): update branchSha (branch="renovate/slf4j.version")
DEBUG: getBranchPr(renovate/slf4j.version) (branch="renovate/slf4j.version")
DEBUG: findPr(renovate/slf4j.version, undefined, open) (branch="renovate/slf4j.version")
DEBUG: findPr(renovate/slf4j.version, undefined, closed) (branch="renovate/slf4j.version")
DEBUG: Found PR #714 (branch="renovate/slf4j.version")
DEBUG: branchExists=false (branch="renovate/slf4j.version")
DEBUG: dependencyDashboardCheck=undefined (branch="renovate/slf4j.version")
DEBUG: Check for closed PR because recreating closed PRs is disabled. (branch="renovate/slf4j.version")
DEBUG: findPr(renovate/slf4j.version, Update slf4j.version to v2.0.10, !open) (branch="renovate/slf4j.version")
DEBUG: Found PR #714 (branch="renovate/slf4j.version")
DEBUG: Found closed PR with current title (branch="renovate/slf4j.version")
DEBUG: Returning PR from cache (branch="renovate/slf4j.version")
DEBUG: Closed PR already exists. Skipping branch. (branch="renovate/slf4j.version")
{
  "prTitle": "Update slf4j.version to v2.0.10"
}

DEBUG: Getting comments for #714 (branch="renovate/slf4j.version")
DEBUG: Found 0 comments (branch="renovate/slf4j.version")
DEBUG: Ensuring comment "Renovate Ignore Notification" in #714 (branch="renovate/slf4j.version")
INFO: Comment added (branch="renovate/slf4j.version")
{
  "issueNo": 714
  "topic": "Renovate Ignore Notification"
}


```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
Please locate the job log when the PRs were closed

**Reply by tmortagne** [reply]
As I explained in the description, there does not seem to be such log. The last log before the close was more than 1h before (and lasted 12s) and the close is immediately followed by the one I pasted.

Here is the slf4j related log in the previous one to be fully exhaustive:

```
DEBUG: syncBranchState() (branch="renovate/slf4j.version")
DEBUG: branch.isUpToDate(): using cached result "true" (branch="renovate/slf4j.version")
DEBUG: getBranchPr(renovate/slf4j.version) (branch="renovate/slf4j.version")
DEBUG: findPr(renovate/slf4j.version, undefined, open) (branch="renovate/slf4j.version")
DEBUG: Found PR #714 (branch="renovate/slf4j.version")
DEBUG: branchExists=true (branch="renovate/slf4j.version")
DEBUG: dependencyDashboardCheck=undefined (branch="renovate/slf4j.version")
DEBUG: PR rebase requested=false (branch="renovate/slf4j.version")
DEBUG: Checking if PR has been edited (branch="renovate/slf4j.version")
DEBUG: branch.isModified(): using cached result "false" (branch="renovate/slf4j.version")
DEBUG: Found existing branch PR (branch="renovate/slf4j.version")
DEBUG: Checking schedule(at any time, null) (branch="renovate/slf4j.version")
DEBUG: No schedule defined (branch="renovate/slf4j.version")
DEBUG: Branch already exists (branch="renovate/slf4j.version")
DEBUG: Skipping behind base branch check due to rebaseWhen=auto (branch="renovate/slf4j.version")
DEBUG: isBranchConflicted(master, renovate/slf4j.version) (branch="renovate/slf4j.version")
DEBUG: branch.isConflicted(): using cached result "false" (branch="renovate/slf4j.version")
DEBUG: Branch does not need rebasing (branch="renovate/slf4j.version")
DEBUG: Using reuseExistingBranch: true (branch="renovate/slf4j.version")
DEBUG: Checking if we can automerge branch (branch="renovate/slf4j.version")
DEBUG: mergeStatus=no automerge (branch="renovate/slf4j.version")
DEBUG: Ensuring PR (branch="renovate/slf4j.version")
DEBUG: There are 0 errors and 0 warnings (branch="renovate/slf4j.version")
DEBUG: getBranchPr(renovate/slf4j.version) (branch="renovate/slf4j.version")
DEBUG: findPr(renovate/slf4j.version, undefined, open) (branch="renovate/slf4j.version")
DEBUG: Found PR #714 (branch="renovate/slf4j.version")
DEBUG: getPrCache() (branch="renovate/slf4j.version")
DEBUG: Found existing PR (branch="renovate/slf4j.version")
DEBUG: PR cache matches and no PR changes in last 24hrs, so skipping PR body check (branch="renovate/slf4j.version")
DEBUG: PR is not configured for automerge (branch="renovate/slf4j.version")
```

**Reply by rarkins** [reply]
When renovate closes PRs it appends "- autoclosed" to the title. If it hasn't done that, and you can't find a log of it running at the time of PR closing, then there's a good chance that Renovate didn't close it.

**Reply by tmortagne** [reply]
@rarkins not sure what more you tell you. https://github.com/xwiki/xwiki-commons/pull/714 is in public access, so you can see yourself that the PR was closed by user @renovate-bot and that it did not append any  "- autoclosed" to the title.

Is there any way for me to give you read access to https://developer.mend.io/github/xwiki/xwiki-commons ?

**Reply by rarkins** [reply]
I can get access next week when I'm back from vacation.

The fact that it's Forking Renovate and not the main app is very meaningful though. The forking approach does some "acrobatics" with forking, although I don't think any logic has changed recently.

**Reply by viceice** [reply]
maybe the source branches were deleted (by accident) and GitHub auto closed the PR's?

**Reply by tmortagne** [reply]
> maybe the source branches were deleted (by accident) and GitHub auto closed the PR's?

Well, that's not really what GitHub suggests happened on https://github.com/xwiki/xwiki-commons/pull/714.

**Reply by nabeelsaabna** [reply]
Hi,
I have check all logs near the time of PR deletion and the `Renovate Ignore Notification` comment and didn't see any indication that renovate closed the PR (based on the logs).

I tested what viceice said and it's correct

> maybe the source branches were deleted (by accident) and GitHub auto closed the PR's?

if the source branch `renovate-bot:renovate/slf4j.version` gets deleted, the PR will be marked as closed by the owner of the forked repository (renovate-bot)

Now the question remains, who/why/how did the `renovate-bot:renovate/slf4j.version` got deleted ?!!

**Reply by tmortagne** [reply]
> who/why/how did the renovate-bot:renovate/slf4j.version got deleted ?!!

Note that we are not talking only about this one, I just gave this one as example, but there were 20 of those. All the remaining renovate PR on https://github.com/xwiki/xwiki-commons/ and https://github.com/xwiki/xwiki-platform/ were close at the same times in two distinct groups. Here is one example for platform: https://github.com/xwiki/xwiki-platform/pull/2719 (renamed so that renovate re-create a new one). It really does not feel like a human removing a branch with some bad manipulation.

Is it possible that those forks were re-created for some reason ?

**Reply by rarkins** [reply]
@Nabeel I wonder if there's a second repo installed into forking renovate with same origin repo? GitHub allows only one fork of a repo per account so maybe a second account has installed a fork of this?

**Reply by nabeelsaabna** [reply]
actually yes, it's owned by @tmortagne (repo https://github.com/tmortagne/xwiki-commons)

found the job: `018cc26d-8b45-7b03-baaa-b9c73eca5bb5`

with message:
```json
{
	"name": "renovate",
	"level": 30,
	"repository": "tmortagne/xwiki-commons",
	"branch": "renovate/slf4j.version",
	"msg": "Deleting orphan branch",
	"time": "2024-01-01T08:58:47.854Z"
}
```
it matches both time and activity

**Reply by tmortagne** [reply]
What could have triggered this ? I definitely did not enable renovate on this repo 3 days ago (and the same problem happen twice before), or I would have thought of this as a cause right away. I definitely don't need renovate to be enabled on this fork (I actually don't need the fork at all anymore, so I'm simply going to delete it), I probably enabled it on my whole account to try it before using it on https://github.com/xwiki.

That being said, I find it a bit strange that renovate is mixing the repository where it was enabled with some potentially totally unrelated repository from which it was initially forked. I'm afraid anyone with a xwiki fork could potentially cause this again if all it takes is enabling renovate on your own fork.

**Reply by nabeelsaabna** [reply]
@rarkins we are sending `"forkProcessing": "disabled"` so most likely a Renovate OSS bug
checking to try and understand where is the problem

**Reply by rarkins** [reply]
So it successfully skipped/blocked the fork but it pruned the branches before exiting?

**Reply by rarkins** [reply]
@tmortagne Renovate periodically retries disabled repositories just in case they were disabled by mistake or the reason for disabling them has changed

**Reply by nabeelsaabna** [reply]
`"forkProcessing": "disabled"` but we don't see the message: `Repository config enables forks - continuing`
which means that line `#70` is reached
https://github.com/renovatebot/renovate/blob/72fe67dd27bc539f26efc671c20938ca29bc5099/lib/workers/repository/init/apis.ts#L63-L74

and I do see the message: `"msg": "Repository is a fork and not manually configured - skipping - did you want to run with flag --include-forks?"`
and after that immediately we reach the message: `"msg": "Removing any stale branches",` 
and then:
```json
	{
		"name": "renovate",
		"repository": "tmortagne/xwiki-commons",
		"branchList": [],
		"renovateBranches": [
			"renovate/checkstyle.version",
			"renovate/jackson.version",
			"renovate/log4j.version",
			"renovate/major-3-maven-remote-resources-plugin.version",
			"renovate/maven.compiler.version",
			"renovate/org.codehaus.plexus-plexus-archiver-4.x",
			"renovate/org.codehaus.plexus-plexus-classworlds-2.x",
			"renovate/org.codehaus.plexus-plexus-interactivity-api-1.x",
			"renovate/org.codehaus.plexus-plexus-interpolation-1.x",
			"renovate/org.infinispan-infinispan-core-14.x",
			"renovate/org.javassist-javassist-3.x",
			"renovate/plexus-containers.version",
			"renovate/slf4j.version",
			"renovate/webdrivermanager.version"
		],
		"msg": "Branch lists",
		"time": "2024-01-01T08:58:40.023Z",
	}
```

**Reply by rarkins** [reply]
So we should set pruneStaleBranches=false to be safe?

**Reply by nabeelsaabna** [reply]
ok, it's a quick-fix workaround that we can implement now.
code fix can come later..

**Reply by tmortagne** [reply]
Thanks guys for the intense discussion on this issue and what seems like a fix :)

**Reply by nabeelsaabna** [reply]
I added `pruneStaleBranches=false` to Forking Renovate App
feel free to play with triggering jobs on both repo (original and fork) to make sure it's stable

**Reply by tmortagne** [reply]
>  feel free to play with triggering jobs on both repo (original and fork) to make sure it's stable

I deleted the old fork that was in my account, unfortunately.

**Comment by tmortagne** [comment]
Looks like this bug is coming back, see for example https://github.com/xwiki-contrib/application-scripting-documentation/pull/3 which just been closed for no reason. The trigger does not seems to be the same as last time as this repository does not have any close (other than renovate itself) this time according to https://github.com/xwiki-contrib/application-scripting-documentation/forks.

According to the renovate log, it seems the cause is that renovate does not really make a difference between failing to get an artifact and getting a 404 (the repository it's using is very slow today from time to time):

```
DEBUG: GET https://nexus.xwiki.org/nexus/content/repositories/public/org/webjars/require-css/maven-metadata.xml = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=403 retryCount=0, duration=1520)
DEBUG: Dependency lookup unauthorized. Please add authentication with a hostRule for https://nexus.xwiki.org/nexus/content/repositories/public/org/webjars/require-css/maven-metadata.xml
DEBUG: Content is not found for Maven url
{
  "url": "https://nexus.xwiki.org/nexus/content/repositories/public/org/webjars/require-css/maven-metadata.xml"
}

DEBUG: Failed to look up maven package org.webjars:require-css
{
  "dependency": "org.webjars:require-css"
  "packageFile": "scripting-documentation-ui/pom.xml"
}
```

```
{
            "currentValue": "0.1.7",
            "currentVersion": "0.1.7",
            "datasource": "maven",
            "depName": "org.webjars:require-css",
            "depType": "runtime",
            "fileReplacePosition": 2332,
            "fixedVersion": "0.1.7",
            "homepage": "http://webjars.org",
            "isSingleVersion": true,
            "packageName": "org.webjars:require-css",
            "packageScope": "org.webjars",
            "registryUrl": "https://nexus.xwiki.org/nexus/content/repositories/public",
            "registryUrls": [
              "https://repo.maven.apache.org/maven2"
            ],
            "sourceUrl": "https://github.com/webjars/require-css",
            "versioning": "maven",
            "warnings": [
              {
                "topic": "org.webjars:require-css",
                "message": "Failed to look up maven package org.webjars:require-css"
              }
            ],
            "updates": []
          },
```

```
DEBUG: findPr(renovate/org.webjars-require-css-0.x, undefined, open)
DEBUG: Found PR #3
DEBUG: branch.isModified(): using git to calculate
DEBUG: branch.isModified() = false
DEBUG: setCachedModifiedResult(): Branch cache not present
INFO: Autoclosing PR
{
  "branchName": "renovate/org.webjars-require-css-0.x"
  "prNo": 3
  "prTitle": "Update dependency org.webjars:require-css to v0.1.10"
}
```

**Reply by tmortagne** [reply]
I was hoping to find an option to tell renovate to never close a PR just because some dependency cannot be found anymore, but I could not spot anything like this on https://docs.renovatebot.com/configuration-options/.
